articles/speech-recognition-in-javascript/index.md

```diff
@@ -0,0 +1,288 @@
+
+In the starter code, there is a `<p>` element with an ID called status that says `Listening...`. It's been hidden by setting the display property of the element to none using CSS.
+
+Let's set it to `display: block` when the speech recognition starts.
+
+```JavaScript
+speechRecognition.onstart = () => {
+  document.querySelector("#status").style.display = "block";
+  console.log("Speech Recognition Started");
+};
+```
+
+#### onEnd
+This event is triggered when speech recognition is ended by the user. Let's pass a callback function that will hide the status `<p>` element in the webpage.
+
+Let's set it to `display: none` when the speech recognition starts.
+
+```JavaScript
+speechRecognition.onend = () => {
+  document.querySelector("#status").style.display = "none";
+  console.log("Speech Recognition Ended");
+};
+```
+
+#### onError
+This event is triggered when there is some sort of error in speech recognition. Let's pass a callback function that will hide the status `<p>` element in the webpage.
+
+Let's set it to `display: none` when the speech recognition starts.
+
+```JavaScript
+speechRecognition.onError = () => {
+  document.querySelector("#status").style.display = "none";
+  console.log("Speech Recognition Error");
+};
+```
+
+#### onResult
+This event is triggered when the `speechRecognition` object has some results from the recognition. It will contain the final results and interim results. Let's pass a callback function that will set the results to the respective `<span>` inside the transcript box.
+
+This is the HTML code for the transcript box on the web page. The interim results span is colored in a different color to differentiate between the interim results and the final results.
+
+```html
+<div class="p-3" style="border: 1px solid gray; height: 300px; border-radius: 8px;">
+  <span id="final" class="text-light"></span>
+  <span id="interim" class="text-secondary"></span>
+</div>
+```
+
+We need to set the interim results to the `span` with the ID interim and the final results to the `span` with the ID final.
+
+The result event will pass an `event` object to the callback function. This object will contain the results in the form of an array. Each element in the array will have a property called `isFinal` denoting whether that item is an interim result or a final result.
+
+Let's declare a variable for the final transcript outside the callback function and a variable for the interim transcript inside the callback function. 
+
+```JavaScript
+let final_transcript = "";
+
+speechRecognition.onresult = (event) => {
+  let interim_transcript = "";
+};
+``` 
+
+Now let's build a string from the results array. We should run it through a loop and add the result item to the final transcript if the result item is final. If not, we should add it to the interim results string.
+
+```JavaScript
```

Comment:
These code snippets have been explained pretty well, but wouldn't it be better if we can have some helpful code comments for this block as well as for the final `speechRecognition.js` file?

Reply:
I agree. Will add them ASAP.

Reply:
Done.